src/libcore/slice.rs

```diff
@@ -295,22 +295,23 @@ impl<T> SliceExt for [T] {
     fn binary_search_by<F>(&self, mut f: F) -> Result<usize, usize> where
         F: FnMut(&T) -> Ordering
     {
-        let mut base : usize = 0;
-        let mut lim : usize = self.len();
+        let mut base = 0usize;
+        let mut s = self;
 
-        while lim != 0 {
-            let ix = base + (lim >> 1);
-            match f(&self[ix]) {
-                Equal => return Ok(ix),
+        loop {
+            let (head, tail) = s.split_at(s.len() >> 1);
+            if tail.is_empty() {
+                return Err(base)
+            }
+            match f(&tail[0]) {
+                Equal => return Ok(base + head.len()),
```

Comment:
Ideally it is not needed (i.e. LLVM will know how to best reorder the code), but moving the check for `Equal` to the last match arm would likely make sure that it is only checked once.
Assuming no clever optimisations and that the comparison results is `Greater` `G` times and `Less` `L` times before you get the desired `Equal` result, the original order will perform `2*G+2*L+1` comparisons in the `match`, while moving `Equal` to the last arm would result in `L+2*G+1` comparisons.